File: code/15.go
/*
 * @Author: lmio
 * @Date: 2023-02-21 16:50:11
 * @LastEditTime: 2023-02-21 18:45:01
 * @FilePath: /leetcode/code/15.go
 * @Description:15.三数之和
 */
package code

import "sort"

func ThreeSum(nums []int) [][]int {
	sort.Ints(nums)
	length := len(nums)
	res := [][]int{}
	nextval := func(index *int) {
		val := nums[*index]
		for *index < length && nums[*index] == val {
			*index++
		}
	}
	prevval := func(index *int) {
		val := nums[*index]
		for *index > 0 && nums[*index] == val {
			*index--
		}
	}
	for i := 0; i < length-2 && nums[i] <= 0; nextval(&i) {
		j := i + 1
		k := length - 1
		for j < k {
			sum := nums[i] + nums[j] + nums[k]
			if sum == 0 {
				res = append(res, []int{nums[i], nums[j], nums[k]})
				nextval(&j)
				prevval(&k)
			} else if sum < 0 {
				j++
			} else {
				k--
			}
		}
	}
	return res
}

File: code_cpp/15.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int> &nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < nums.size() - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            if (nums[i] > 0) {
                break;
            }
            if (nums[i] + nums[nums.size() - 1] + nums[nums.size() - 2] < 0) {
                continue;
            }
            int l = i + 1, r = nums.size() - 1;
            while (l < r) {
                auto sum = nums[i] + nums[l] + nums[r];
                if (sum == 0) {
                    res.push_back({nums[i], nums[l], nums[r]});
                    l++;
                    r--;
                    while (l < r && nums[l] == nums[l - 1])
                        l++;
                    while (l < r && nums[r] == nums[r + 1])
                        r--;
                } else if (sum < 0) {
                    l++;
                } else {
                    r--;
                }
            }
        }
        return res;
    }
};